pkg/kubelet/images/puller.go

```diff
@@ -40,22 +41,26 @@ type imagePuller interface {
 var _, _ imagePuller = &parallelImagePuller{}, &serialImagePuller{}
 
 type parallelImagePuller struct {
-	imageService kubecontainer.ImageService
-	tokens       chan struct{}
+	imageService      kubecontainer.ImageService
+	tokens            chan struct{}
+	pullingImageIDSet sync.Map
 }
 
 func newParallelImagePuller(imageService kubecontainer.ImageService, maxParallelImagePulls *int32) imagePuller {
 	if maxParallelImagePulls == nil || *maxParallelImagePulls < 1 {
-		return &parallelImagePuller{imageService, nil}
+		return &parallelImagePuller{imageService, nil, sync.Map{}}
 	}
-	return &parallelImagePuller{imageService, make(chan struct{}, *maxParallelImagePulls)}
+	return &parallelImagePuller{imageService, make(chan struct{}, *maxParallelImagePulls), sync.Map{}}
 }
 
 func (pip *parallelImagePuller) pullImage(ctx context.Context, spec kubecontainer.ImageSpec, pullSecrets []v1.Secret, pullChan chan<- pullResult, podSandboxConfig *runtimeapi.PodSandboxConfig) {
 	go func() {
-		if pip.tokens != nil {
+		if _, loaded := pip.pullingImageIDSet.LoadOrStore(spec.Image, struct{}{}); pip.tokens != nil && !loaded {
 			pip.tokens <- struct{}{}
-			defer func() { <-pip.tokens }()
+			defer func() {
+				pip.pullingImageIDSet.Delete(spec.Image)
```

Comment:
Using only the image name, we cannot confirm the uniqueness of the image. Please refer to the existing pattern in #125192.
```
func pullKey(spec kubecontainer.ImageSpec) string {
	res := []string{}
	for _, annotation := range spec.Annotations {
		res = append(res, annotation.Name+"-"+annotation.Value)
	}
	sort.Strings(res)
	if spec.RuntimeHandler != "" {
		res = append([]string{spec.RuntimeHandler}, res...)
	}
	if spec.Image != "" {
		res = append([]string{spec.Image}, res...)
	}
	return strings.Join(res, "-")
}
```

Reply:
looks like pullKey() contains more informations into key。
according to https://github.com/kubernetes/kubernetes/blob/f8d5b2074c9b3a1629ef085b9e17a5370f393732/pkg/kubelet/images/image_manager.go#L121

Reply:
Using the image's `Annotation` and `RuntimeHandler` as the unique key helps us distinguish between multiple image versions in Windows environment.
ref: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/4216-image-pull-per-runtime-class

However, I also don't believe that we need to include all `Annotations` as part of the key. `Annotations` like `kubernetes.io/config.seen` will make each pod's image always unique.

Do you think we should select some valuable Annotations as part of the unique key? For example, `kubernetes.io/runtimehandler`?

@saschagrunert @SergeyKanzhelev

Reply:
This is an interesting KEP. Sorry I didn't read it before.

Reply:
It seems I didn't express clearly.

The `RuntimeHandler` and the `Annotation` named `kubernetes.io/runtimehandler` are not used to distinguish **container runtimes** but rather the **container runtime class**.

You can find valuable information about **container runtime class** at the following link: https://kubernetes.io/docs/concepts/containers/runtime-class/.

The goal of KEP-4216 is to distinguish between different versions of the same image name through `RuntimeClassName` (which will be populated in the `RuntimeHandler` field of the `ImageSpec`).

ref: https://github.com/kubernetes/kubernetes/blob/dbc2b0a5c7acc349ea71a14e49913661eaf708d2/pkg/kubelet/kuberuntime/kuberuntime_container.go#L172-L187

Reply:
Sorry, I misunderstood before.🙏 We can select different runtimehandlers provide 
from CRI to complete image pulling by configuring pod's runtime class.
From this perspective, since implementation logic of different runtime classes may different, it seems necessary for us to distinguish different runtime classes while queue up.

Reply:
/cc @haircommander, do you have any suggestions regarding https://github.com/kubernetes/kubernetes/pull/126382#discussion_r1700499445? Do we need to add any annotations other than `kubernetes.io/runtimehandler` as part of the unique key?

Reply:
I really don't want to put any assumptions in kubelet on how CRI interprets the annotations. If we need to filter out some - I'd prefer to filter them out in `PullImage` call as well.

Alternative is to sequence image pulls that are "almost" the same, not pull them in-parallel. This alternative also makes an assumption about container runtime, but not a very critical assumption.